### PR TITLE
 feat(security)!: Implementation of JWT authentication ADR

### PR DESCRIPTION
**Pull request by bnevis-i**
BREAKING CHANGE: Requires JWT authentication for all inbound
requests except for /api/v2/ping URL. Removes support for
Kong reverse proxy.  In place of Kong, uses NGINX
proxy auth module and introduces new security-prox-auth service.
Changes secrets-config proxy adduser/deluser commands to create
Vault users instead of Kong user.  Changes secrets-config proxy tls
command to write TLS certificate to docker volume instead of Kong.
Removes security-proxy-setup go binary and replaces with shell
script to create default TLS token.

Signed-off-by: Bryon Nevis <bryon.nevis@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [X] I am not introducing a new dependency (add notes below if you are)
- [X] I have added unit tests for the new feature or bug fix (if not, why?)
- [X] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

### CODE REVIEW

**Comment by bnevis-i** [comment]
@farshidtz Blocked by https://github.com/canonical/edgex-snap-testing/pull/163

**Comment by farshidtz** [comment]
> @farshidtz To be unblocked by [canonical/edgex-snap-testing#163](https://github.com/canonical/edgex-snap-testing/pull/163)

Thanks, looking into it. But please keep in mind that the snap tests don't block PR merges.

File: internal/security/config/command/proxy/shared/proxyuser.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
+//
+// Copyright (c) 2020-2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0'
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/shared/proxyuser.go:4]:
```suggestion
// SPDX-License-Identifier: Apache-2.0
```

File: internal/security/config/command/proxy/shared/proxyuser.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+type ProxyUserCommon struct {
+	loggingClient     logger.LoggingClient
+	configuration     *secretStoreConfig.ConfigurationStruct
+	fileOpener        fileioperformer.FileIoPerformer
+	httpCaller        internal.HttpCaller
+	secretStoreClient secrets.SecretStoreClient
+}
+
+// NewVaultBase sets up everything needed to instantiate a SecretStoreClient
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/shared/proxyuser.go:48]:
```suggestion
// NewProxyUserCommon sets up everything needed to instantiate a SecretStoreClient
```

File: internal/security/config/command/proxy/shared/proxyuser.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	token, err := tokenLoader.Load(tokenPath)
+	if err != nil {
+		return "", revokeFunc, err
+	}
+
+	return token, revokeFunc, nil
+
+}
+
+// loadRootToken regenerates a temporary root token from Vault keyshares
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/shared/proxyuser.go:115]:
```suggestion
// LoadRootToken regenerates a temporary root token from Vault keyshares
```

**Review by jim-wang-intel — CHANGES_REQUESTED** [review]
some minor changes

File: cmd/security-bootstrapper/entrypoint-scripts/nginx_wait_install.sh [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+#  SPDX-License-Identifier: Apache-2.0
+#  ----------------------------------------------------------------------------------
+
+# This is customized entrypoint script for Kong.
+# In particular, it waits for the security-bootstrapper's ReadyToRunPort and Postgres db ready to roll
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-bootstrapper/entrypoint-scripts/nginx_wait_install.sh]: fixed

File: cmd/security-bootstrapper/entrypoint-scripts/nginx_wait_install.sh [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+set -e
+
+# env settings are populated from env files of docker-compose
+
+echo "Script for waiting security bootstrapping on Kong"
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-bootstrapper/entrypoint-scripts/nginx_wait_install.sh]: fixed.

File: cmd/security-bootstrapper/entrypoint-scripts/nginx_wait_install.sh [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+# gating on the ready-to-run port
+echo "$(date) Executing waitFor with waiting on tcp://${STAGEGATE_BOOTSTRAPPER_HOST}:${STAGEGATE_READY_TORUNPORT}"
+/edgex-init/security-bootstrapper --configDir=/edgex-init/res waitFor \
+  -uri tcp://"${STAGEGATE_BOOTSTRAPPER_HOST}":"${STAGEGATE_READY_TORUNPORT}" \
+  -timeout "${STAGEGATE_WAITFOR_TIMEOUT}"
+
+
+echo "$(date) Generating default config ..."
+
+test -f /etc/nginx/templates/edgex-custom-rewrites.inc.template || cat <<'EOH' > /etc/nginx/templates/edgex-custom-rewrites.inc.template
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-bootstrapper/entrypoint-scripts/nginx_wait_install.sh]: fixed with long-form if..else

File: cmd/security-bootstrapper/entrypoint-scripts/proxy_setup_wait_install.sh [hunk 7]
```diff
@@ -33,25 +33,5 @@ echo "$(date) Executing waitFor for ${PROXY_SETUP_HOST} with waiting on \
   -uri tcp://"${STAGEGATE_BOOTSTRAPPER_HOST}":"${STAGEGATE_READY_TORUNPORT}" \
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-bootstrapper/entrypoint-scripts/proxy_setup_wait_install.sh:32]: fixed comment above.

File: cmd/security-proxy-setup/Dockerfile [hunk 8]
```diff
@@ -27,21 +27,20 @@ COPY go.mod vendor* ./
...
-RUN make cmd/security-proxy-setup/security-proxy-setup cmd/secrets-config/secrets-config
+RUN make cmd/secrets-config/secrets-config
 
 FROM alpine:3.16
 
-RUN apk add --update --no-cache curl dumb-init
+RUN apk add --update --no-cache dumb-init openssl
 
 LABEL license='SPDX-License-Identifier: Apache-2.0' \
       copyright='Copyright (c) 2019: Dell Technologies, Inc.'
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-proxy-setup/Dockerfile]: added

File: internal/security/config/command/proxy/shared/proxyuser.go (hunk 1 above)
Comment (bnevis-i) [reply internal/security/config/command/proxy/shared/proxyuser.go]: fixed typo

File: internal/security/config/command/proxy/shared/proxyuser.go (hunk 2 above)
Comment (bnevis-i) [reply internal/security/config/command/proxy/shared/proxyuser.go]: fixed

File: internal/security/config/command/proxy/shared/proxyuser.go (hunk 3 above)
Comment (bnevis-i) [reply internal/security/config/command/proxy/shared/proxyuser.go]: fixed

File: internal/security/config/command/proxy/tls/command.go [hunk 9]
```diff
@@ -1,60 +1,61 @@
 //
-// Copyright (c) 2020 Intel Corporation
+// Copyright (c) 2020-2023 Intel Corporation
 //
 // SPDX-License-Identifier: Apache-2.0'
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/tls/command.go:4]: just saw this:
```suggestion
// SPDX-License-Identifier: Apache-2.0
```

File: internal/security/config/bootstraphandler.go [hunk 10]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2020 Intel Corporation
+// Copyright (c) 2020-2023 Intel Corporation
 //
 // SPDX-License-Identifier: Apache-2.0'
```
Comment (jim-wang-intel) [inline internal/security/config/bootstraphandler.go:4]: just saw this:
```suggestion
// SPDX-License-Identifier: Apache-2.0
```

File: internal/security/config/command/help/command.go [hunk 11]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2020 Intel Corporation
+// Copyright (c) 2020-2023 Intel Corporation
 //
 // SPDX-License-Identifier: Apache-2.0'
```
Comment (jim-wang-intel) [inline internal/security/config/command/help/command.go:4]: just saw this:
```suggestion
// SPDX-License-Identifier: Apache-2.0
```

File: internal/security/config/command/proxy/adduser/command.go [hunk 12]
```diff
@@ -1,287 +1,107 @@
 //
-// Copyright (c) 2020 Intel Corporation
+// Copyright (c) 2020-2023 Intel Corporation
 //
 // SPDX-License-Identifier: Apache-2.0'
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/adduser/command.go:4]: just saw this:
```suggestion
// SPDX-License-Identifier: Apache-2.0
```

File: internal/security/config/command/proxy/command.go [hunk 13]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2020 Intel Corporation
+// Copyright (c) 2020-2023 Intel Corporation
 //
 // SPDX-License-Identifier: Apache-2.0'
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/command.go:4]: sorry, just saw this:
```suggestion
// SPDX-License-Identifier: Apache-2.0
```

File: internal/security/config/command/proxy/deluser/command.go [hunk 14]
```diff
@@ -1,5 +1,5 @@
 //
-// Copyright (c) 2020 Intel Corporation
+// Copyright (c) 2020-2023 Intel Corporation
 //
 // SPDX-License-Identifier: Apache-2.0'
```
Comment (jim-wang-intel) [inline internal/security/config/command/proxy/deluser/command.go:4]:
```suggestion
// SPDX-License-Identifier: Apache-2.0
```

File: internal/security/config/command/proxy/tls/command.go (hunk 9 above)
Comment (bnevis-i) [reply internal/security/config/command/proxy/tls/command.go]: fixed in rebase

File: internal/security/config/bootstraphandler.go (hunk 10 above)
Comment (bnevis-i) [reply internal/security/config/bootstraphandler.go]: fixed in rebase

File: internal/security/config/command/proxy/adduser/command.go (hunk 12 above)
Comment (bnevis-i) [reply internal/security/config/command/proxy/adduser/command.go]: fixed in rebase

File: internal/security/config/command/help/command.go (hunk 11 above)
Comment (bnevis-i) [reply internal/security/config/command/help/command.go]: fixed in rebase

File: internal/security/config/command/proxy/command.go (hunk 13 above)
Comment (bnevis-i) [reply internal/security/config/command/proxy/command.go]: fixed in rebase

File: internal/security/config/command/proxy/deluser/command.go (hunk 14 above)
Comment (bnevis-i) [reply internal/security/config/command/proxy/deluser/command.go]: fixed in rebase

**Review by jim-wang-intel — APPROVED** [review]
LGTM

**Review by farshidtz — COMMENTED** [review]
Already merged but I'd like to suggest making the CLI argument naming conventions consistent:

```
$ edgexfoundry.secrets-config proxy adduser -h
Usage of adduser:
  -configDir string
    	
  -jwtTTL string
    	JWT created by vault identity provider lasts this long (default "1h")
  -tokenTTL string
    	Vault token created as a result of vault login lasts this long (default "1h")
  -useRootToken
    	Set to true to TokenFile in config points to a resp-init.json instead of a service token
  -user string
    	Username of the user to add
level=ERROR ts=2023-03-10T11:55:04.114755814Z app=secrets-config source=bootstraphandler.go:70 msg="unable to parse command: -h: flag: help requested"
```
Ok - all camelCase

```
$ edgexfoundry.secrets-config proxy deluser -h
Usage of deluser:
  -configDir string
    	
  -useRootToken
    	Set to true to TokenFile in config points to a resp-init.json instead of a service token
  -user string
    	Username of the user to delete
level=ERROR ts=2023-03-10T11:55:10.795005034Z app=secrets-config source=bootstraphandler.go:70 msg="Unable to parse command: -h: flag: help requested"
```
Ok - all camelCase

```
$ edgexfoundry.secrets-config proxy tls -h
Usage of tls:
  -certfilename string
    	Filename of certificate file (on target) (default "nginx.crt")
  -configDir string
    	
  -incert string
    	Path to PEM-encoded leaf certificate
  -inkey string
    	Path to PEM-encoded private key
  -keyfilename string
    	Filename of private key file (on target (default "nginx.key")
  -targetfolder string
    	Path to TLS key file (default "/etc/ssl/nginx")
level=ERROR ts=2023-03-10T11:55:14.547332083Z app=secrets-config source=bootstraphandler.go:70 msg="unable to parse command: -h: flag: help requested"
```
Not OK.

- `-certfilename` -> `-certFilename`
- `-incert` -> `-inCert`
- `-inkey` -> `inKey`
- `-keyfilename` -> `-keyFilename`
- `-targetfolder` -> `-targetFolder`

Also, I don't know why it is printing error on `-h`.

@bnevis-i I'll open an issue if necessary.

**Comment by farshidtz** [comment]
Related PRs:
- https://github.com/edgexfoundry/edgex-docs/pull/981
- https://github.com/edgexfoundry/edgex-compose/pull/345

**Comment by bnevis-i** [comment]
@farshidtz  See #4433 for fixes described in https://github.com/edgexfoundry/edgex-go/pull/4244#pullrequestreview-1334725750

**Pull request merged**